osc/conf.py

```diff
@@ -519,10 +533,141 @@ def _rewind_request(self, req):
+                if line == "-----BEGIN OPENSSH PRIVATE KEY-----":
+                    return True
+            return False
+
+        def list_ssh_dir_keys(self):
+            sshdir = os.path.expanduser('~/.ssh')
+            keys_in_home_ssh = {}
+            for keyfile in os.listdir(sshdir):
+                if keyfile.endswith(".pub"):
+                    continue
+                keyfile_path = os.path.join(sshdir, keyfile)
+                if not self.is_ssh_private_keyfile(keyfile_path):
+                    continue
+                cmd = ["ssh-keygen", "-lf", keyfile_path]
+                proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+                stdout, _ = proc.communicate()
+                if proc.returncode == 0:
+                    fingerprint = stdout.strip()
+                    if fingerprint:
+                        keys_in_home_ssh[fingerprint] = keyfile_path
+            return keys_in_home_ssh
+
+        def guess_keyfile(self):
+            keys_in_agent = self.list_ssh_agent_keys()
+            if keys_in_agent:
+                keys_in_home_ssh = self.list_ssh_dir_keys()
+                for fingerprint in keys_in_agent:
+                    if fingerprint in keys_in_home_ssh:
+                        return keys_in_home_ssh[fingerprint]
+            sshdir = os.path.expanduser('~/.ssh')
+            keyfiles = ('id_ed25519', 'id_rsa')
+            for keyfile in keyfiles:
+                keyfile_path = os.path.join(sshdir, keyfile)
+                if os.path.isfile(keyfile_path):
+                    return keyfile_path
+            raise oscerr.OscIOError(None, 'could not guess ssh identity keyfile')
+
+        def ssh_sign(self, data, namespace, keyfile=None):
+            try:
+                data = bytes(data, 'utf-8')
+            except:
+                pass
+            if not keyfile:
+                keyfile = self.guess_keyfile()
+            else:
+                if '/' not in keyfile:
+                    keyfile = '~/.ssh/' + keyfile
+                keyfile = os.path.expanduser(keyfile)
+
+            cmd = ['ssh-keygen', '-Y', 'sign', '-f', keyfile, '-n', namespace, '-q']
+            proc = subprocess.Popen(cmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE)
+            stdout, _ = proc.communicate(data)
+            if proc.returncode:
+                raise oscerr.OscIOError(None, 'ssh-keygen signature creation failed: %d' % proc.returncode)
+
+            signature = decode_it(stdout)
+            match = re.match(r"\A-----BEGIN SSH SIGNATURE-----\n(.*)\n-----END SSH SIGNATURE-----", signature, re.S)
+            if not match:
+                raise oscerr.OscIOError(None, 'could not extract ssh signature')
+            return base64.b64decode(match.group(1))
+
+        def get_authorization(self, req, chal):
+            realm = chal.get('realm', '')
+            now = int(time.time())
+            sigdata = "(created): %d" % now
```

Comment:
Hi there, I'm currently implementing this logic for [OSC-Tiny](https://github.com/crazyscientist/osc-tiny) and noticed that the substring `(created)` would have been available from the response header, too.

If you don't mind my asking, any particular reason for hard-coding this substring?

Reply:
Thanks for sharing the background info.

>  the client sends what it actually used (which can be a superset).

After looking into the specs of [HTTP Message Signatures](https://tools.ietf.org/id/draft-cavage-http-signatures-12.html), my understanding would be similar, but the server keeps responding with 401, if I try to use additional data in the signature string.

But this would be a(n academic) topic for the server-side implementation.